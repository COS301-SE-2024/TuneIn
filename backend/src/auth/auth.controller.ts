import {
    Controller,
    Post,
    Body,
    HttpException,
    HttpStatus,
    Get
} from "@nestjs/common";
import {
	AuthService,
	CognitoDecodedToken,
	JWTPayload,
	RegisterBody,
	LoginBody,
} from "./auth.service";
import { ApiBody, ApiOperation, ApiResponse, ApiTags } from "@nestjs/swagger";
import { UsersService } from "src/modules/users/users.service";
import * as PrismaTypes from "@prisma/client";

@ApiTags('auth')
@Controller("auth")
export class AuthController {
	constructor(
		private readonly authService: AuthService,
		private readonly usersService: UsersService,
	) {}

	@Post("login")
	@ApiTags("auth")
	@ApiOperation({ summary: "Login in the API using Cognito" })
	@ApiBody({ type: LoginBody })
	@ApiResponse({
		status: 201,
		description: "The record has been successfully created.",
		type: String,
	})
	@ApiResponse({ status: 403, description: "Forbidden." })
	async login(@Body() loginInfo: LoginBody) {
		/*
		const users = await this.authService.listUsers();
		console.log(users);
		console.log(users.Users);
		if (!users) {
			throw new HttpException(
				"Invalid credentials. Could not create user. AuthError01",
				HttpStatus.UNAUTHORIZED,
			);
		}

        if (!users.Users || users.Users.length === 0) {
            throw new HttpException(
                "Invalid credentials. Could not create user. AuthError02",
                HttpStatus.UNAUTHORIZED,
            );
        }

        // Match the email address given with the email in the UserPool
        let userMatch = null;
        let userEmail = "";
        for (let i = 0; i < users.Users.length; i++) {
            if (!users.Users[i] || users.Users[i] === undefined) {
                continue;
            }
            console.log("users.Users[i]", users.Users[i]);
            const attrs = users.Users[i].Attributes;
            console.log("attrs", attrs);
            if (!attrs || attrs === undefined || attrs.length === 0) {
                continue;
            }

            // If "email" is the Name of an attribute in users.Users[i].Attributes
            if (attrs.find((attribute) => attribute.Name === "email")) {
                const attr = attrs.find((attribute) => attribute.Name === "email");
                if (!attr || attr === undefined) {
                    continue;
                }
                // If the userCognitoSub matches the UserSub in the UserPool
                if (users.Users[i].Username === authInfo.userCognitoSub && attr.Value) {
                    userMatch = users.Users[i];
                    userEmail = attr.Value;
                    break;
                }
            }
        }

        if (userMatch === null || !userMatch) {
            throw new HttpException(
                "Invalid credentials. Could not create user. AuthError03",
                HttpStatus.UNAUTHORIZED,
            );
        }

        // Add users to table
        const successful: boolean = await this.authService.createUser(
            authInfo.username,
            userEmail,
            authInfo.userCognitoSub,
        );

        if (!successful) {
            throw new HttpException(
                "Invalid credentials. Could not create user. AuthError04",
                HttpStatus.UNAUTHORIZED,
            );
        }

        const payload = {
            sub: authInfo.userCognitoSub,
            username: authInfo.username,
            email: userEmail,
        };

        // Generate JWT token using payload
        const token: string = await this.authService.generateJWT(payload);

		//return the JWT as a string
		return { token: token };
	}

	@Post("register")
	@ApiTags("auth")
	@ApiOperation({ summary: "Register a new user in the API using Cognito" })
	@ApiBody({ type: RegisterBody })
	@ApiResponse({
		status: 201,
		description: "The record has been successfully created.",
		type: RegisterBody,
	})
	@ApiResponse({ status: 403, description: "Forbidden." })
	async register(@Body() registerInfo: RegisterBody) {
		const successful: boolean = await this.authService.createUser(
			registerInfo.username,
			registerInfo.email,
			registerInfo.userCognitoSub,
		);

		if (!successful) {
			throw new HttpException(
				"Invalid credentials. Could not create user. AuthRegisterError01",
				HttpStatus.UNAUTHORIZED,
			);
		}

		throw new HttpException("Successfully created user.", HttpStatus.CREATED);
	}

    // TODO: Add a POST method to refresh an expired JWT token
    // TODO: Add a POST method to logout a user
}
