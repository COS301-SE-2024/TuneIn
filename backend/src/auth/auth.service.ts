import { Injectable, UnauthorizedException } from "@nestjs/common";
import * as AWS from "aws-sdk";
import { ConfigService } from "@nestjs/config";
import { PrismaService } from "../../prisma/prisma.service";
import { Prisma } from "@prisma/client";
import * as jwt from "jsonwebtoken";
import { CognitoJwtVerifier } from "aws-jwt-verify";
import { ApiProperty } from "@nestjs/swagger";
//import { CreateUserDto } from "src/modules/users/dto/create-user.dto";

export type CognitoDecodedToken = {
	sub: string;
	iss: string;
	client_id: string;
	origin_jti: string;
	event_id?: string;
	token_use: string;
	scope?: string;
	auth_time: number;
	exp: number;
	iat: number;
	jti: string;
	username: string;
};

export type JWTPayload = {
	id: string;
	email: string;
	username: string;
};

export class RegisterBody {
	@ApiProperty()
	username: string;

	@ApiProperty()
	userCognitoSub: string;

	@ApiProperty()
	email: string;
}

export class LoginBody {
	@ApiProperty()
	token: string;
}

@Injectable()
export class AuthService {
	private cognitoIdentityServiceProvider: AWS.CognitoIdentityServiceProvider;
	private accessKeyId: string;
	private secretAccessKey: string;
	private userPoolId: string;
	private clientId: string;

	constructor(
		private configService: ConfigService,
		private prisma: PrismaService,
	) {
		// Set the AWS credentials
		const accessKeyId = this.configService.get<string>("AWS_ACCESS_KEY_ID");
		if (!accessKeyId) {
			throw new Error("Missing AWS_ACCESS_KEY_ID");
		}
		this.accessKeyId = accessKeyId;

		const secretAccessKey = this.configService.get<string>(
			"AWS_SECRET_ACCESS_KEY",
		);
		if (!secretAccessKey) {
			throw new Error("Missing AWS_SECRET_ACCESS_KEY");
		}
		this.secretAccessKey = secretAccessKey;

		const userPoolId = this.configService.get<string>(
			"AWS_COGNITO_USER_POOL_ID",
		);
		if (!userPoolId) {
			throw new Error("Missing AWS_COGNITO_USER_POOL_ID");
		}
		this.userPoolId = userPoolId;

		const clientId = this.configService.get<string>("AWS_COGNITO_CLIENT_ID");
		if (!clientId) {
			throw new Error("Missing AWS_COGNITO_CLIENT_ID");
		}
		this.clientId = clientId;

		this.cognitoIdentityServiceProvider =
			new AWS.CognitoIdentityServiceProvider({
				region: "af-south-1",
				accessKeyId: this.accessKeyId,
				secretAccessKey: this.secretAccessKey,
			});
	}

	async validateUser(username: string, password: string): Promise<any> {
		if (!this.userPoolId || !this.clientId) {
			throw new Error("Missing Cognito configuration");
		}

		const params: AWS.CognitoIdentityServiceProvider.AdminInitiateAuthRequest =
			{
				AuthFlow: "ADMIN_NO_SRP_AUTH",
				ClientId: this.clientId,
				UserPoolId: this.userPoolId,
				AuthParameters: {
					USERNAME: username,
					PASSWORD: password,
				},
			};

		try {
			const authResponse = await this.cognitoIdentityServiceProvider
				.adminInitiateAuth(params)
				.promise();
			if (authResponse.AuthenticationResult) {
				const user = await this.cognitoIdentityServiceProvider
					.adminGetUser({
						UserPoolId: this.userPoolId,
						Username: username,
					})
					.promise();

				return {
					username: user.Username,
					attributes: user.UserAttributes,
					authenticationResult: authResponse.AuthenticationResult,
				};
			}
		} catch (error) {
			throw new UnauthorizedException("Invalid credentials");
		}
	}

	// eslint-disable-next-line prettier/prettier
  	async listUsers(): Promise<AWS.CognitoIdentityServiceProvider.ListUsersResponse> {
		const params = {
			UserPoolId: this.userPoolId,
		};

		try {
			const responsePromise = await this.cognitoIdentityServiceProvider
				.listUsers(params)
				.promise();
			const response = await responsePromise;
			return response;
		} catch (error) {
			throw new UnauthorizedException("Error listing users");
		}
	}

	async createUser(
		username: string,
		email: string,
		userID: string,
	): Promise<boolean> {
		const user: Prisma.usersCreateInput = {
			username: username,
			email: email,
			user_id: userID,
		};
		const existingUser = await this.prisma.users.findUnique({
			where: { user_id: userID },
		});
		if (existingUser) {
			return true;
		}
		try {
			const response = await this.prisma.users.create({ data: user });
			console.log(response);
		} catch (err) {
			console.log(err);
			return false;
		}
		return true;
	}

	async generateJWT(payload: JWTPayload): Promise<string> {
		const secretKey = this.configService.get<string>("JWT_SECRET_KEY");
		const expiresIn = this.configService.get<string>("JWT_EXPIRATION_TIME");

		if (!secretKey || secretKey === undefined || secretKey === "") {
			throw new Error("Missing JWT secret key");
		}

		if (!expiresIn || expiresIn === undefined || expiresIn === "") {
			throw new Error("Missing JWT expiration time");
		}

		const token = jwt.sign(payload, secretKey, { expiresIn });
		return token;
	}

	/*
	async getUserInfo(jwt_token: string): Promise<any> {
		const secretKey = this.configService.get<string>("JWT_SECRET_KEY");
		const expiresIn = this.configService.get<string>("JWT_EXPIRATION_TIME");

		if (!secretKey || secretKey === undefined || secretKey === "") {
			throw new Error("Missing JWT secret key");
		}

		if (!expiresIn || expiresIn === undefined || expiresIn === "") {
			throw new Error("Missing JWT expiration time");
		}

		const decoded = jwt.verify(jwt_token, secretKey);
		const userID = decoded.sub;
		if (!userID) {
			throw new Error("Invalid JWT token");
		}
		if (typeof userID !== "string") {
			throw new Error("Invalid JWT token");
		}

		const user = await this.prisma.users.findUnique({
			where: { user_id: userID },
		});
		return user;
	}
	*/

	// this funciton will be passed a Request object from the NestJS controller eg:
	/*
	getRoomInfo(@Request() req: any, @Param("roomID") roomID: string): RoomDto {
		return this.roomsService.getRoomInfo(roomID);
	}
	*/
	getUserInfo(req: any): JWTPayload {
		console.log("req", req);
		const result = req.user as JWTPayload;
		console.log(result);
		if (!result) {
			throw new UnauthorizedException(
				"No user found in JWT token. Please log in again",
			);
		}
		if (!result.id) {
			throw new UnauthorizedException(
				"No user ID found in JWT token. Please log in again",
			);
		}
		return result;
	}

	async decodeAndVerifyCognitoJWT(
		jwt_token: string,
	): Promise<CognitoDecodedToken> {
		const verifier = CognitoJwtVerifier.create({
			userPoolId: this.userPoolId,
			tokenUse: "access",
			clientId: this.clientId,
		});

		try {
			const payload = await verifier.verify(jwt_token);
			const result: CognitoDecodedToken = payload as CognitoDecodedToken;
			return result;
		} catch (error) {
			throw new UnauthorizedException("Invalid JWT token");
		}
	}
}
