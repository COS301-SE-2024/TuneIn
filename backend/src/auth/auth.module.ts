import { Module } from "@nestjs/common";
import { AuthService } from "./auth.service";
import { AuthController } from "./auth.controller";
import { PassportModule } from "@nestjs/passport";
import { JwtModule } from "@nestjs/jwt";
import { LocalStrategy } from "./local.strategy";
import { JwtStrategy } from "./jwt.strategy";
import { ConfigModule } from "@nestjs/config";
import { PrismaService } from "../../prisma/prisma.service";
import { PrismaModule } from "../../prisma/prisma.module";
import { UsersService } from "src/modules/users/users.service";
import { DbUtilsService } from "src/modules/db-utils/db-utils.service";
import { DtoGenService } from "src/modules/dto-gen/dto-gen.service";
import { SpotifyAuthService } from './spotify/spotify.service';
import { SpotifyAuthController } from './spotify/spotify.controller';
import { SpotifyAuthModule } from './spotify/spotify.module';

@Module({
	imports: [
		PassportModule,
		JwtModule.register({
			secret: process.env.JWT_SECRET || "your_jwt_secret",
			signOptions: { expiresIn: "2h" },
		}),
		ConfigModule.forRoot(), // Ensure ConfigModule is imported to access environment variables
		PrismaModule, SpotifyAuthModule,
	],
	providers: [
		AuthService,
		LocalStrategy,
		JwtStrategy,
		PrismaService,
		DtoGenService,
		DbUtilsService,
		UsersService,
		SpotifyAuthService,
	],
	controllers: [AuthController, SpotifyAuthController],
})
export class AuthModule {}
