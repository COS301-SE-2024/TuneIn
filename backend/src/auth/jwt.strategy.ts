import { Strategy, ExtractJwt } from "passport-jwt";
import { PassportStrategy } from "@nestjs/passport";
import { Injectable } from "@nestjs/common";
import { AuthService } from "./auth.service";

@Injectable()
export class JwtStrategy extends PassportStrategy(Strategy) {
	constructor(private authService: AuthService) {
		super({
			jwtFromRequest: ExtractJwt.fromAuthHeaderAsBearerToken(),
			ignoreExpiration: false,
			secretOrKey: "your_jwt_secret",
		});
	}

	async validate(payload: any) {
		return { userId: payload.sub, username: payload.username };
	}
}
