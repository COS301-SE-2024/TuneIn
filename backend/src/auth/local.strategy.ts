import { Strategy } from "passport-local";
import { PassportStrategy } from "@nestjs/passport";
import { Injectable, UnauthorizedException } from "@nestjs/common";
import { AuthService } from "./auth.service";
import {
	AdminInitiateAuthCommandInput,
	AttributeType,
	AuthenticationResultType,
	CognitoIdentityProvider,
	ListUsersCommandOutput,
} from "@aws-sdk/client-cognito-identity-provider";

@Injectable()
export class LocalStrategy extends PassportStrategy(Strategy) {
	constructor(private authService: AuthService) {
		super();
	}

	async validate(
		username: string,
		password: string,
	): Promise<{
		username: string;
		attributes: AttributeType[];
		authenticationResult: AuthenticationResultType;
	}> {
		const user = await this.authService.validateUser(username, password);
		return user;
	}
}
