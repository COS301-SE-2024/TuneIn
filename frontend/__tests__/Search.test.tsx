import React from "react";
import { render, fireEvent, act, waitFor } from "@testing-library/react-native";
import Search from "../app/screens/Search"; // Adjust the path as needed
import { useNavigation } from "expo-router";
import axios from "axios";
// import AsyncStorage from "@react-native-async-storage/async-storage";
import auth from "../app/services/AuthManagement";
// import Fuse from "fuse.js";

// Mocking modules
jest.mock("expo-router", () => ({
	useNavigation: jest.fn(),
}));

jest.mock("../app/components/rooms/RoomCardWidget", () => {
	const MockRoomCardWidget = (props: Record<string, unknown>) => (
		<div {...props} />
	);
	MockRoomCardWidget.displayName = "MockRoomCardWidget";
	return MockRoomCardWidget;
});

jest.mock("../app/components/UserItem", () => {
	const MockUserItem = (props: Record<string, unknown>) => <div {...props} />;
	MockUserItem.displayName = "MockUserItem";
	return MockUserItem;
});

jest.mock("../app/components/NavBar", () => {
	const MockNavBar = () => <div />;
	MockNavBar.displayName = "MockNavBar";
	return MockNavBar;
});

jest.mock("axios");

// jest.mock('fuse.js', () => {
// 	return jest.fn().mockImplementation(() => {
// 		return {
// 			search: jest.fn(),
// 		};
// 	});
// });

jest.mock("../app/services/AuthManagement", () => ({
	__esModule: true,
	default: {
		getToken: jest.fn(), // Mock getToken method
	},
}));

const roomMock = [
	{
		creator: {
			profile_name: "Farmer23",
			userID: "01ece2d8-e091-7023-c1f2-d3399faa7071",
			username: "farmer 345",
			profile_picture_url:
				"https://tunein-nest-bucket.s3.af-south-1.amazonaws.com/2024-06-23T13:23:20.848Z-image.jpeg",
			followers: { count: 0, data: [] },
			following: { count: 0, data: [] },
			links: { count: 0, data: [] },
			bio: "Music enthusiast who loves exploring new tunes across various genres. Always on the lookout for fresh beats and hidden gems!",
			current_song: {
				title: "",
				artists: [],
				cover: "",
				start_time: "2024-08-11T17:22:28.862Z",
			},
			fav_genres: { count: 0, data: [] },
			fav_songs: { count: 0, data: [] },
			fav_rooms: { count: 0, data: [] },
			recent_rooms: { count: 0, data: [] },
		},
		roomID: "66bb6bf7-25be-45af-bc38-7e7e258797b8",
		participant_count: 0,
		room_name: "chill vibes",
		description:
			"A room for relaxing and enjoying soothing music. Join us to unwind and chill with your favorite tunes.",
		is_temporary: false,
		is_private: false,
		is_scheduled: false,
		start_date: "2024-08-11T17:22:28.862Z",
		end_date: "2024-08-11T17:22:28.862Z",
		language: "English",
		has_explicit_content: true,
		has_nsfw_content: false,
		room_image: "https://ik.imagekit.io/ikmedia/backlit.jpg",
		current_song: {
			title: "",
			artists: [],
			cover: "",
			start_time: "2024-08-11T17:22:28.862Z",
		},
		tags: ["explicit"],
	},
];

const userMock = [
	{
		id: "2",
		type: "user",
		name: "User 1",
		userData: {
			id: "1",
			profile_picture_url:
				"https://wallpapers-clan.com/wp-content/uploads/2023/11/marvel-iron-man-in-destroyed-suit-desktop-wallpaper-preview.jpg",
			profile_name: "User 1",
			username: "user1",
		},
	},
	{
		id: "4",
		type: "user",
		name: "User 2",
		userData: {
			id: "2",
			profile_picture_url:
				"https://wallpapers-clan.com/wp-content/uploads/2023/11/marvel-iron-man-in-destroyed-suit-desktop-wallpaper-preview.jpg",
			profile_name: "User 2",
			username: "user2",
		},
	},
];

const uHistDtoMock = [
	{
		search_term: "nothing",
		search_time: "2024-07-19T09:03:19.651Z",
		url: "/search/users?q=nothing",
	},
];

describe("Search Component", () => {
	beforeEach(() => {
		// Clear mocks before each test
		(useNavigation as jest.Mock).mockClear();
		(axios.get as jest.Mock).mockClear();
		(auth.getToken as jest.Mock).mockReturnValue("token");
	});

	it("should render the header with a title and back button", () => {
		(axios.get as jest.Mock)
			.mockResolvedValueOnce({ data: roomMock })
			.mockResolvedValueOnce({ data: uHistDtoMock })
			.mockResolvedValueOnce({ data: ["jazz", "rock"] })
			.mockResolvedValueOnce({ data: uHistDtoMock });
		const { getByText, getByTestId } = render(<Search />);
		expect(getByText("Search")).toBeTruthy();
		expect(getByTestId("back-button")).toBeTruthy();
	});

	it("should handle search input changes", () => {
		(axios.get as jest.Mock)
			.mockResolvedValueOnce({ data: roomMock })
			.mockResolvedValueOnce({ data: uHistDtoMock })
			.mockResolvedValueOnce({ data: ["jazz", "rock"] })
			.mockResolvedValue({ data: uHistDtoMock });

		const { getByPlaceholderText } = render(<Search />);
		const searchInput = getByPlaceholderText("Search...");
		fireEvent.changeText(searchInput, "Room 1");
		expect(searchInput.props.value).toBe("Room 1");
	});

	it("should toggle more filters", async () => {
		// Mock search result
		(axios.get as jest.Mock)
			.mockResolvedValueOnce({ data: roomMock })
			.mockResolvedValueOnce({ data: uHistDtoMock })
			.mockResolvedValueOnce({ data: ["jazz", "rock"] })
			.mockResolvedValue({ data: uHistDtoMock });

		// Render the page component (this triggers the mock usage)
		const { getByText, getByTestId, queryByTestId } = render(<Search />);

		// Press the button to show more filters
		fireEvent.press(getByText("View More Filters"));

		// Check if the additional filters are shown
		await waitFor(() => {
			expect(getByTestId("host-toggle")).toBeTruthy();
			expect(getByTestId("room-count-toggle")).toBeTruthy();
		});

		// Press the button to hide more filters
		fireEvent.press(getByText("View Less Filters"));

		await waitFor(() => {
			// Check if the additional filters are hidden
			expect(queryByTestId("host-toggle")).toBeNull();
			expect(queryByTestId("room-count-toggle")).toBeNull();
		});
	});

	it("should handle explicit filter switch", async () => {
		(axios.get as jest.Mock)
			.mockResolvedValueOnce({ data: roomMock })
			.mockResolvedValueOnce({ data: uHistDtoMock })
			.mockResolvedValueOnce({ data: ["jazz", "rock"] })
			.mockResolvedValue({ data: uHistDtoMock });
		const { getByTestId, getByText } = render(<Search />);
		fireEvent.press(getByText("View More Filters"));
		const explicitSwitch = getByTestId("explicit-switch");

		// Simulate toggling the switch
		fireEvent(explicitSwitch, "valueChange", true);

		// Verify the switch value has changed
		await waitFor(() => {
			expect(explicitSwitch.props.value).toBe(true);
		});
	});

	it("should handle nsfw filter switch", async () => {
		(axios.get as jest.Mock)
			.mockResolvedValueOnce({ data: roomMock })
			.mockResolvedValueOnce({ data: uHistDtoMock })
			.mockResolvedValueOnce({ data: ["jazz", "rock"] })
			.mockResolvedValue({ data: uHistDtoMock });
		const { getByTestId, getByText } = render(<Search />);
		fireEvent.press(getByText("View More Filters"));
		const nsfwSwitch = getByTestId("nsfw-switch");

		// Simulate toggling the switch
		fireEvent(nsfwSwitch, "valueChange", true);

		// Verify the switch value has changed
		await waitFor(() => {
			expect(nsfwSwitch.props.value).toBe(true);
		});
	});

	it("should handle temp filter switch", async () => {
		(axios.get as jest.Mock)
			.mockResolvedValueOnce({ data: roomMock })
			.mockResolvedValueOnce({ data: uHistDtoMock })
			.mockResolvedValueOnce({ data: ["jazz", "rock"] })
			.mockResolvedValue({ data: uHistDtoMock });
		const { getByTestId, getByText } = render(<Search />);
		fireEvent.press(getByText("View More Filters"));
		const tempSwitch = getByTestId("temp-switch");

		// Simulate toggling the switch
		fireEvent(tempSwitch, "valueChange", true);

		// Verify the switch value has changed
		await waitFor(() => {
			expect(tempSwitch.props.value).toBe(true);
		});
	});

	it("should handle priv filter switch", () => {
		(axios.get as jest.Mock)
			.mockResolvedValueOnce({ data: roomMock })
			.mockResolvedValueOnce({ data: uHistDtoMock })
			.mockResolvedValueOnce({ data: ["jazz", "rock"] })
			.mockResolvedValue({ data: uHistDtoMock });
		const { getByTestId, getByText } = render(<Search />);
		fireEvent.press(getByText("View More Filters"));
		const privSwitch = getByTestId("priv-switch");

		// Simulate toggling the switch
		fireEvent(privSwitch, "valueChange", true);

		// Verify the switch value has changed
		expect(privSwitch.props.value).toBe(true);
	});

	it("should handle scheduled filter switch", async () => {
		(axios.get as jest.Mock)
			.mockResolvedValueOnce({ data: roomMock })
			.mockResolvedValueOnce({ data: uHistDtoMock })
			.mockResolvedValueOnce({ data: ["jazz", "rock"] })
			.mockResolvedValue({ data: uHistDtoMock });
		const { getByTestId, getByText } = render(<Search />);
		fireEvent.press(getByText("View More Filters"));
		const scheduledSwitch = getByTestId("scheduled-switch");

		// Simulate toggling the switch
		fireEvent(scheduledSwitch, "valueChange", true);

		// Verify the switch value has changed
		await waitFor(() => {
			expect(scheduledSwitch.props.value).toBe(true);
		});
	});

	it("should search with all room filters", async () => {
		(axios.get as jest.Mock)
			.mockResolvedValueOnce({ data: roomMock })
			.mockResolvedValueOnce({ data: uHistDtoMock })
			.mockResolvedValueOnce({ data: ["jazz", "rock"] })
			.mockResolvedValueOnce({ data: uHistDtoMock })
			.mockResolvedValueOnce({ data: roomMock })
			.mockResolvedValue({ data: uHistDtoMock });
		const { getByPlaceholderText, getByTestId } = render(<Search />);
		fireEvent.press(getByTestId("toggle-filters-button"));

		// Simulate filter switches
		fireEvent(getByTestId("explicit-switch"), "valueChange", true);
		fireEvent(getByTestId("nsfw-switch"), "valueChange", true);
		fireEvent(getByTestId("temp-switch"), "valueChange", true);
		fireEvent(getByTestId("priv-switch"), "valueChange", true);
		fireEvent(getByTestId("scheduled-switch"), "valueChange", true);

		await waitFor(async () => {
			const searchInput = getByPlaceholderText("Search...");
			fireEvent.changeText(searchInput, "Room 1");
			fireEvent.press(getByTestId("search-button"));
		});
	});

	it("should search with no room filters", async () => {
		(axios.get as jest.Mock)
			.mockResolvedValueOnce({ data: roomMock })
			.mockResolvedValueOnce({ data: uHistDtoMock })
			.mockResolvedValueOnce({ data: ["jazz", "rock"] })
			.mockResolvedValueOnce({ data: [] })
			.mockResolvedValue({ data: uHistDtoMock });

		(axios.get as jest.Mock).mockResolvedValueOnce({ data: roomMock });
		const { getByPlaceholderText, getByTestId, getByText } = render(<Search />);
		fireEvent.press(getByTestId("toggle-filters-button"));

		await waitFor(
			async () => {
				const searchInput = getByPlaceholderText("Search...");
				fireEvent.changeText(searchInput, "nothing");
				fireEvent.press(getByTestId("search-button"));
			},
			{ timeout: 20000 },
		);

		// await waitFor(async () => {
		// 	expect(getByTestId("search-button")).toBe("nothing");
		// });
	}, 30000);
});
