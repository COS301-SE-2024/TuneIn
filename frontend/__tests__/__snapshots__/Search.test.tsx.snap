// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search Component should render correctly 1`] = `
<View
  style={
    {
      "flex": 1,
      "paddingHorizontal": 30,
      "paddingTop": 30,
    }
  }
>
  <View
    style={
      {
        "alignItems": "center",
        "flexDirection": "row",
        "justifyContent": "center",
        "marginBottom": 20,
      }
    }
  >
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "opacity": 1,
        }
      }
      testID="back-button"
    >
      <Text />
    </View>
    <Text
      style={
        {
          "color": "#333",
          "flex": 1,
          "fontSize": 24,
          "fontWeight": "bold",
          "textAlign": "center",
        }
      }
    >
      Search   
    </Text>
  </View>
  <View
    style={
      {
        "alignItems": "center",
        "borderColor": "#ccc",
        "borderRadius": 56,
        "borderWidth": 1,
        "flexDirection": "row",
        "marginBottom": 20,
        "paddingHorizontal": 10,
      }
    }
  >
    <TextInput
      onChangeText={[Function]}
      placeholder="Search..."
      style={
        {
          "flex": 1,
          "height": 40,
        }
      }
      value=""
    />
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "marginLeft": 10,
          "opacity": 1,
        }
      }
      testID="search-button"
    >
      <Text />
    </View>
  </View>
  <View
    style={
      {
        "flexDirection": "row",
        "justifyContent": "space-between",
        "marginBottom": 20,
      }
    }
  >
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "backgroundColor": "#08BDBD",
          "borderColor": "#08BDBD",
          "borderRadius": 7,
          "borderWidth": 1,
          "elevation": 5,
          "opacity": 1,
          "paddingHorizontal": 20,
          "paddingVertical": 10,
          "shadowColor": "#000",
          "shadowOffset": {
            "height": 4,
            "width": 0,
          },
          "shadowOpacity": 0.25,
          "shadowRadius": 5.84,
        }
      }
    >
      <Text
        style={
          {
            "color": "#333",
            "fontWeight": "bold",
          }
        }
      >
        All
      </Text>
    </View>
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "borderColor": "#ccc",
          "borderRadius": 7,
          "borderWidth": 1,
          "opacity": 1,
          "paddingHorizontal": 20,
          "paddingVertical": 10,
        }
      }
    >
      <Text
        style={
          {
            "color": "#333",
            "fontWeight": "bold",
          }
        }
      >
        Rooms
      </Text>
    </View>
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "borderColor": "#ccc",
          "borderRadius": 7,
          "borderWidth": 1,
          "opacity": 1,
          "paddingHorizontal": 20,
          "paddingVertical": 10,
        }
      }
    >
      <Text
        style={
          {
            "color": "#333",
            "fontWeight": "bold",
          }
        }
      >
        Users
      </Text>
    </View>
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "borderColor": "#ccc",
          "borderRadius": 7,
          "borderWidth": 1,
          "opacity": 1,
          "paddingHorizontal": 20,
          "paddingVertical": 10,
        }
      }
      testID="filter-button"
    >
      <Text />
    </View>
  </View>
  <View
    style={
      {
        "flexDirection": "row",
        "flexWrap": "wrap",
        "marginBottom": 20,
      }
    }
  />
  <RCTScrollView
    contentContainerStyle={
      {
        "paddingBottom": 20,
      }
    }
    data={[]}
    getItem={[Function]}
    getItemCount={[Function]}
    keyExtractor={[Function]}
    onContentSizeChange={[Function]}
    onLayout={[Function]}
    onMomentumScrollBegin={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={0.0001}
    stickyHeaderIndices={[]}
    testID="scroll-view"
    viewabilityConfigCallbackPairs={[]}
  >
    <View />
  </RCTScrollView>
  <View
    collapsable={false}
    style={
      {
        "bottom": 0,
        "left": 0,
        "position": "absolute",
        "right": 0,
        "transform": [
          {
            "translateY": 0,
          },
        ],
        "zIndex": 10,
      }
    }
  >
    <div />
  </View>
</View>
`;
